File: data/multilang_comparison/program_cpp_vector/src/func38.c
#include "program_cpp_vector.h" 
array_t* func38(array_t_param* vars, const unsigned long PATH0, int loopsFactor) {
   size_t pCounter = vars->size;
   if(PATH0 & 1) {
      array_t* array77;
      if (pCounter > 0) {
         array77 = vars->data[--pCounter];
         array77->refC++;
         DEBUG_COPY(array77->id);
      } else {
         array77 = (array_t*)malloc(sizeof(array_t));
         array77->size = 416;
         array77->refC = 1;
         array77->id = 77;
         array77->data = (unsigned int*)malloc(array77->size*sizeof(unsigned int));
         memset(array77->data, 0, array77->size*sizeof(unsigned int));
         DEBUG_NEW(array77->id);
      }
      unsigned int loop25 = 0;
      unsigned int loopLimit25 = (rand()%loopsFactor)/4 + 1;
      for(; loop25 < loopLimit25; loop25++) {
         for (int i = 0; i < array77->size; i++) {
            array77->data[i]--;
         }
         for (int i = 0; i < array77->size; i++) {
            if (array77->data[i] == 24) { 
               return array77;
            }
         }
      }
      array_t_param params0;
      params0.size = 1;
      params0.data = (array_t**)malloc(params0.size*sizeof(array_t*));
      params0.data[0] = array77;
      array_t* array78 = func46(&params0, get_path(), loopsFactor);
      DEBUG_RETURN(array78->id);
      free(params0.data);
      array78->refC--;
      if(array78->refC == 0) {
         free(array78->data);
         free(array78);
         DEBUG_FREE(array78->id);
      }
      array77->refC--;
      if(array77->refC == 0) {
         free(array77->data);
         free(array77);
         DEBUG_FREE(array77->id);
      }
   }
   else {
      array_t_param params0;
      params0.size = 0;
      params0.data = (array_t**)malloc(params0.size*sizeof(array_t*));
      array_t* array79 = func40(&params0, loopsFactor);
      DEBUG_RETURN(array79->id);
      free(params0.data);
      array79->refC--;
      if(array79->refC == 0) {
         free(array79->data);
         free(array79);
         DEBUG_FREE(array79->id);
      }
   }
   array_t* array80;
   if (pCounter > 0) {
      array80 = vars->data[--pCounter];
      array80->refC++;
      DEBUG_COPY(array80->id);
   } else {
      array80 = (array_t*)malloc(sizeof(array_t));
      array80->size = 36;
      array80->refC = 1;
      array80->id = 80;
      array80->data = (unsigned int*)malloc(array80->size*sizeof(unsigned int));
      memset(array80->data, 0, array80->size*sizeof(unsigned int));
      DEBUG_NEW(array80->id);
   }
   return array80;
}

File: data/multilang_comparison/program_nim/src/func38.nim
proc func38*(vars: ptr ArrayTParam, PATH0: uint64, loopsFactor: cint): ptr ArrayT =
   var pCounter = vars.size
   if (PATH0 and 1'u64) != 0:
      discard
      var array139: ptr ArrayT
      if pCounter > 0:
         dec pCounter
         array139 = vars.data[pCounter]
         inc array139.refC
         debugCopy(array139.id)
      else:
         array139 = cast[ptr ArrayT](alloc(sizeof(ArrayT)))
         array139.size = 255
         array139.refC = 1
         array139.id = 139
         array139.data = cast[ptr UncheckedArray[uint32]](alloc(array139.size * sizeof(uint32)))
         zeroMem(array139.data, array139.size * sizeof(uint32))
         debugNew(array139.id)
      let loopLimit43 = (50) div (4) + 1
      for loop43 in 0 ..< loopLimit43:
         discard
         for i in 0 ..< array139.size:
            dec array139.data[i]
         for i in 0 ..< array139.size:
            if array139.data[i] == 79:
               return array139
      var params0: ArrayTParam
      params0.size = 1
      params0.data = cast[ptr UncheckedArray[ptr ArrayT]](alloc(params0.size * sizeof(ptr ArrayT)))
      params0.data[0] = array139
      var array140: ptr ArrayT = func46(addr params0, getPath(), loopsFactor)
      debugReturn(array140.id)
      dealloc(params0.data)
      dec array140.refC
      if array140.refC == 0:
         dealloc(array140.data)
         dealloc(array140)
         debugFree(array140.id)
      dec array139.refC
      if array139.refC == 0:
         dealloc(array139.data)
         dealloc(array139)
         debugFree(array139.id)
   else:
      discard
      var params0: ArrayTParam
      params0.size = 0
      params0.data = cast[ptr UncheckedArray[ptr ArrayT]](alloc(params0.size * sizeof(ptr ArrayT)))
      var array141: ptr ArrayT = func40(addr params0, getPath(), loopsFactor)
      debugReturn(array141.id)
      dealloc(params0.data)
      dec array141.refC
      if array141.refC == 0:
         dealloc(array141.data)
         dealloc(array141)
         debugFree(array141.id)
   var array142: ptr ArrayT
   if pCounter > 0:
      dec pCounter
      array142 = vars.data[pCounter]
      inc array142.refC
      debugCopy(array142.id)
   else:
      array142 = cast[ptr ArrayT](alloc(sizeof(ArrayT)))
      array142.size = 567
      array142.refC = 1
      array142.id = 142
      array142.data = cast[ptr UncheckedArray[uint32]](alloc(array142.size * sizeof(uint32)))
      zeroMem(array142.data, array142.size * sizeof(uint32))
      debugNew(array142.id)
   return array142

